pyanaconda/modules/subscription/subscription.py

```diff
@@ -26,3 +38,183 @@ class SubscriptionService(KickstartService):
+
+        self.set_system_purpose_data(system_purpose_data)
+
+    def setup_kickstart(self, data):
+        """Return the kickstart string."""
+        log.debug("Generating kickstart data...")
+
+        # system purpose
+        data.syspurpose.role = self.system_purpose_data.role
+        data.syspurpose.sla = self.system_purpose_data.sla
+        data.syspurpose.usage = self.system_purpose_data.usage
+        data.syspurpose.addons = self.system_purpose_data.addons
+
+        return str(data)
+
+    # system purpose configuration
+
+    def _load_valid_system_purpose_values(self):
+        """Load lists of valid roles, SLAs and usage types.
+
+        About role/sla/validity:
+        - an older installation image might have older list of valid fields,
+          missing fields that have become valid after the image has been released
+        - fields that have been valid in the past might be dropped in the future
+        - there is no list of valid addons
+
+        Due to this we need to take into account that the listing might not always be
+        comprehensive and that we need to allow what might on a first glance look like
+        invalid values to be written to the target system.
+        """
+        roles, slas, usage_types = system_purpose.get_valid_fields()
+        self._valid_roles = roles
+        self._valid_slas = slas
+        self._valid_usage_types = usage_types
+
+    def _process_role(self, data):
+        if data.syspurpose.role:
+            role_match = system_purpose.match_field(data.syspurpose.role, self.valid_roles)
+        else:
+            role_match = None
+
+        if role_match:
+            log.info("role value %s from kickstart matched to known valid field %s",
+                     data.syspurpose.role,
+                     role_match)
+            return role_match
+        elif data.syspurpose.role:
+            log.info("using custom role value from kickstart: %s", data.syspurpose.role)
+            return data.syspurpose.role
+        else:
+            return ""
+
+    def _process_sla(self, data):
+        if data.syspurpose.sla:
+            sla_match = system_purpose.match_field(data.syspurpose.sla, self.valid_slas)
+        else:
+            sla_match = None
+
+        if sla_match:
+            log.info("SLA value %s from kickstart matched to known valid field %s",
+                     data.syspurpose.sla,
+                     sla_match)
+            return sla_match
+        elif data.syspurpose.sla:
+            log.info("using custom SLA value from kickstart: %s", data.syspurpose.sla)
+            return data.syspurpose.sla
+        else:
+            return ""
+
+    def _process_usage(self, data):
+        if data.syspurpose.usage:
+            usage_match = system_purpose.match_field(data.syspurpose.usage,
+                                                     self._valid_usage_types)
+        else:
+            usage_match = None
```

Comment:
What about returning `None` if  the argument doesn't exists. You then can remove this from three places to just one.
Or in general, wouldn't it be better to have all the `process_*` private methods as part of the `system_purpose` library. It seems to me that it would make the main service cleaner and avoid repetition.

Reply:
Code was moved to `system_purpose` library.